faw/ci/ci_container_service.py

```diff
@@ -1307,6 +1578,101 @@ async def _get_decisions(self, req):
     return t
 
 
+def start_devmount_watcher(*, config_dir, build_dir, devmounts, faw_docker_id, builder_thread):
+
+    import watchfiles
+
+    class DevMountWatcher(threading.Thread):
+        def __init__(self):
+            super().__init__(daemon=True)
+            self._stop_event = threading.Event()
+
+            # Compute the set of files we need to look for and
+            # map it to the stages we would need to rebuild in
+            # case they change. We also collect a map from files
+            # to devmounts.
+            self._path_stages_map = collections.defaultdict(set)
+            self._path_devmounts_map = collections.defaultdict(set)
+            for name, info in devmounts.items():
+                if not info.valid:
+                    continue
+
+                all_paths = []
+                for p in info.trigger_paths:
+                    full_path = os.path.join(info.mounted_path, p)
+                    paths = glob.glob(full_path, recursive=True)
+                    all_paths.extend(paths)
+
+                for path in all_paths:
+                    for stage_name in info.dependent_stages:
+                        self._path_stages_map[path].add(stage_name)
+
+                    self._path_devmounts_map[path].add(name)
+
+            logging.info(f"Watching {list(self._path_stages_map.keys())}")
+
+        def stop(self):
+            self._stop_event.set()
+
+        def run(self):
+            for changes in watchfiles.watch(*self._path_stages_map.keys(), yield_on_timeout=True):
+                # If we have been asked to stop, do that
+                if self._stop_event.is_set():
+                    return
+
+                # Figure out the set of stages we need to rebuild as well as
+                # devmounts that were affected by this
+                updated_stage_names = set()
+                affected_devmounts = set()
+                for change, path in changes:
+                    # We react to everything other than deletion of the path in question
+                    # TODO: Should we do this?
+                    if change == watchfiles.Change.deleted:
+                        continue
```

Comment:
I believe this would trigger if any of the `triggers` have been deleted. I have added logging for this case for the time being in commit: [0d22404](https://github.com/GaloisInc/FAW/pull/90/commits/0d224042004210af60d5cfd1c5b60312f6461ab0).